File: integration_tests/typescript/src/round-trip.ts
import { deepStrictEqual } from 'assert';

export function checkMatch<T, U>(
  size: (value: T) => number,
  serialize: (dataView: DataView, offset: number, value: T) => number,
  deserialize: (dataView: DataView, offset: number) => [number, U],
  value: T,
): void {
  // eslint-disable-next-line no-console
  console.log('Value to be serialized:', value);

  const valueSize = size(value);
  // eslint-disable-next-line no-console
  console.log('Expected size of the serialized value:', valueSize);

  const arrayBuffer = new ArrayBuffer(valueSize);
  const dataView = new DataView(arrayBuffer);

  const numBytesWritten = serialize(dataView, 0, value);
  deepStrictEqual(numBytesWritten, valueSize);

  // eslint-disable-next-line no-console
  console.log('Bytes from serialization:', arrayBuffer);

  // eslint-disable-next-line no-console
  console.log('Size of the serialized value:', arrayBuffer.byteLength);

  const [numBytesRead, clone] = deserialize(dataView, 0);

  // eslint-disable-next-line no-console
  console.log('Value deserialized from those bytes:', clone);

  deepStrictEqual(numBytesRead, numBytesWritten);
  deepStrictEqual(clone, value);
}

export function checkOk<T, U>(
  size: (value: T) => number,
  serialize: (dataView: DataView, offset: number, value: T) => number,
  deserialize: (dataView: DataView, offset: number) => [number, U],
  value: T,
): void {
  // eslint-disable-next-line no-console
  console.log('Value to be serialized:', value);

  const valueSize = size(value);
  // eslint-disable-next-line no-console
  console.log('Expected size of the serialized value:', valueSize);

  const arrayBuffer = new ArrayBuffer(valueSize);
  const dataView = new DataView(arrayBuffer);

  const numBytesWritten = serialize(dataView, 0, value);
  deepStrictEqual(numBytesWritten, valueSize);

  // eslint-disable-next-line no-console
  console.log('Bytes from serialization:', arrayBuffer);

  // eslint-disable-next-line no-console
  console.log('Size of the serialized value:', arrayBuffer.byteLength);

  const [numBytesRead, clone] = deserialize(dataView, 0);

  // eslint-disable-next-line no-console
  console.log('Value deserialized from those bytes:', clone);

  deepStrictEqual(numBytesRead, numBytesWritten);
}

File: integration_tests/rust/src/round_trip.rs
use {
    crate::types::{Deserialize, Serialize},
    std::{
        fmt::Debug,
        io::{self, Error, ErrorKind},
    },
};

pub fn check_match<T: Debug + Serialize, U: Debug + Deserialize + From<T>>(x: T) -> io::Result<()> {
    println!("Value to be serialized: {:?}", x);

    let size = x.size();
    println!("Expected size of the serialized value: {:?}", size);

    let mut buffer = Vec::<u8>::new();
    x.serialize(&mut buffer)?;
    println!("Bytes from serialization: {:?}", buffer);

    println!("Size of the serialized value: {:?}", buffer.len());
    if buffer.len() != size {
        return Err(Error::new(ErrorKind::Other, "Mismatch!"));
    }

    let mut slice = buffer.as_slice();
    let y = U::deserialize(&mut slice)?;
    println!("Value deserialized from those bytes: {:?}", y);

    if !slice.is_empty() {
        return Err(Error::new(
            ErrorKind::Other,
            "The buffer was not consumed completely!",
        ));
    }

    if format!("{:?}", y) != format!("{:?}", U::from(x)) {
        return Err(Error::new(ErrorKind::Other, "Mismatch!"));
    }

    Ok(())
}

pub fn check_ok<T: Debug + Serialize, U: Debug + Deserialize>(x: &T) -> io::Result<()> {
    println!("Value to be serialized: {:?}", x);

    let size = x.size();
    println!("Expected size of the serialized value: {:?}", size);

    let mut buffer = Vec::<u8>::new();
    x.serialize(&mut buffer)?;
    println!("Bytes from serialization: {:?}", buffer);

    println!("Size of the serialized value: {:?}", buffer.len());
    if buffer.len() != size {
        return Err(Error::new(ErrorKind::Other, "Mismatch!"));
    }

    let mut slice = buffer.as_slice();
    let y = U::deserialize(&mut slice)?;
    println!("Value deserialized from those bytes: {:?}", y);

    if !slice.is_empty() {
        return Err(Error::new(
            ErrorKind::Other,
            "The buffer was not consumed completely!",
        ));
    }

    Ok(())
}
